src/mapping/query/function/flatten.rs

```diff
@@ -0,0 +1,362 @@
+use super::prelude::*;
+use std::collections::btree_map;
+
+/// An iterator to walk over maps allowing us to flatten nested maps to a single level.
+struct MapFlatten<'a> {
+    values: btree_map::Iter<'a, String, Value>,
+    inner: Option<Box<MapFlatten<'a>>>,
+    parent: Option<String>,
+}
+
+impl<'a> MapFlatten<'a> {
+    fn new(values: btree_map::Iter<'a, String, Value>) -> Self {
+        Self {
+            values,
+            inner: None,
+            parent: None,
+        }
+    }
+
+    fn new_from_parent(parent: String, values: btree_map::Iter<'a, String, Value>) -> Self {
+        Self {
+            values,
+            inner: None,
+            parent: Some(parent),
+        }
+    }
+
+    /// Returns the key with the parent prepended.
+    fn new_key(&self, key: &str) -> String {
+        match self.parent {
+            None => key.to_string(),
+            Some(ref parent) => format!("{}.{}", parent, key),
+        }
+    }
+}
+
+impl<'a> std::iter::Iterator for MapFlatten<'a> {
+    type Item = (String, &'a Value);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(ref mut inner) = self.inner {
+            let next = inner.next();
+            if next.is_some() {
+                return next;
+            } else {
+                self.inner = None;
+            }
+        }
+
+        let next = self.values.next();
+        match next {
+            Some((key, Value::Map(value))) => {
+                self.inner = Some(Box::new(MapFlatten::new_from_parent(
+                    self.new_key(key),
+                    value.iter(),
+                )));
+                self.next()
+            }
+            Some((key, value)) => Some((self.new_key(key), value)),
+            None => None,
+        }
+    }
+}
+
+/// Create an iterator that can walk a tree of Array values.
+/// This can be used to flatten the array.
+struct ValueFlatten<'a> {
```

Comment:
Shouldn't this be named `ArrayFlatten`, similar to the above `MapFlatten`?

Reply:
:+1:  Gosh, how did I miss that!?